src/core.jl

```diff
@@ -264,3 +264,46 @@ end
 # Once Base has colon defined here we can replace this
 _colon(I::CartesianIndex{N}, J::CartesianIndex{N}) where N =
     map((i,j) -> i:j, Tuple(I), Tuple(J))
+
+
+"""
+    G, vertex2cartesian = region_adjacency_graph(img, weight_fn, r)
+
+Constructs a region adjacency graph (RAG) from a N-D image `img`. It returns the RAG
+along with a mapping from vertex index in RAG to cartesian index in the image.
+
+`weight_fn` is used to assign weights to the edges using pixel similarity and spatial
+proximity. Zero weight is assigned to edges between any pair of nodes that are more
+than r pixels apart.
+
+    edge_weight = weight_fn(I::CartesianIndex{N}, img[I]::CT, J::CartesianIndex{N}, img[J]::CT) where {CT<:Union{Colorant,Real}, N}
+
+Any graph clustering technique can be used with the constructed RAG to segment the image.
+"""
+
+function region_adjacency_graph(img::AbstractArray{CT,N}, weight_fn::Function, r::Int) where {CT<:Union{Colorant,Real}, N}
```

Comment:
Many 3d images are collected with inhomogenous resolution (e.g., the "mri" example in TestImages has resolution 1mm×1mm×5mm). Might be nice to allow a tuple or CartesianIndex input for `r`? (I.e., directly supplying your `R` below.) Then the `::Int` version could be a convenience method.